incubator/hnc/internal/reconcilers/object_test.go

```diff
@@ -45,18 +45,46 @@ var _ = Describe("Secret", func() {
 		cleanupObjects(ctx)
 	})
 
-	// Exceptions have not been implemented, so we are ignoring the following four tests for now.
+	// Exceptions have not been implemented, so we are ignoring some of the following tests for now.
 	// They should be enabled once exception is done.
-	PIt("should propagate object only to selected namespace using treeSelect", func() {
+	It("should propagate object only to selected namespace using treeSelect", func() {
 		setParent(ctx, barName, fooName)
 		setParent(ctx, bazName, fooName)
 
 		// Create a Role that does NOT propogate to bazName
 		a := map[string]string{"propagate.hnc.x-k8s.io/treeSelect": "!" + bazName}
 		makeObjectWithAnnotation(ctx, "Role", fooName, "testrole", a)
 
-		Eventually(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue())
-		Consistently(hasObject(ctx, "Role", bazName, "testrole")).Should(BeFalse())
+		Eventually(hasObject(ctx, "Role", barName, "testrole")).Should(BeTrue(), "When propagating testrole to %s", barName)
+		Consistently(hasObject(ctx, "Role", bazName, "testrole")).Should(BeFalse(), "When propagating testrole to %s", bazName)
+	})
+
+	It("should propagate object only to selected namespace using multiple treeSelect", func() {
+		setParent(ctx, barName, fooName)
+		setParent(ctx, bazName, fooName)
+
+		// Create a Role that does NOT propogate to bazName
```

Comment:
This comment is out of date. Also, the test might be stronger if `foo` had _three_ children, two of which are excluded but one of which is not.

Finally, all your tests use negation. It would be nice to have a test that _only_ propagated to 2/3 named namespaces. E.g., let's say you had three children (bar, baz, bam). The selectors might be:
* bar (should only propagate to bar)
* !bar (should only propagate to baz/bam)
* bar,baz (should only propagate to bar/baz)
* !bar,!baz (should only propagate to bam)

This looks a lot like a table-driven test - e.g. https://github.com/kubernetes-sigs/multi-tenancy/blob/master/incubator/hnc/internal/validators/hierarchy_test.go#L18. Each test case is a single line and not an entire function. E.g. what I just wrote above could look like:

```
tests := []struct{
  selector string
  want []string
}{
  {selector: barName, want: []string{barName}},
  {selector: "!"+barName, want: []string{bazName, bamName}},
  ...
}
```

And then you write the test _once_ that sets up the object and then verifies its result. Does this make sense?

Reply:
I'm not sure if I understood the structure of this test - does this function still goes under `object_test.go`? Do I also need to set up the test and hierarchy like in the example? I.e.```f := foresttest.Create("-a-") // a <- b; c
h := &Hierarchy{Forest: f}
l := zap.Logger(false)
```

Reply:
No, those three setup lines aren't needed if you're in the reconciler integ tests. And yes, you should keep the tests in the file where they are now, just structure them differently.

You can think of it like this: instead of writing each test individually, you're writing a function to create the tests for you. See the other comment I made for how I _think_ it should work, but TBH I've never written a table test in Ginkgo before - what I wrote these is my best guess.

I'd recommend starting small. See if you can write a dumb table test with two subcases that don't do anything but make sure that they run, e.g.:

```
for i:=0; i <=1; i++ {
  FIt("should verify that i == 0", func() {
    Expect(i == 0)
  })
}
```

That should give you one passing and one failing test. See if that works! And if so, expand from there.

Reply:
done